### PR TITLE
Packaging Modernisation

### PR DESCRIPTION
**Pull request by AA-Turner**
PR for Engelbert Gruber @grubert

### CODE REVIEW

**Comment by chrisjsewell** [comment]
One question on flit: does it properly support editable builds yet, i.e. using `pip install -e .`?
This was my main problem with flit, when developing a package you want editable installs to properly jump around the source code in things like VS Code.
Note, pip now supports editable installs with the setuptools build system and no setup.py necessary

**Comment by AA-Turner** [comment]
I'm using PyCharm & haven't had any problems with `pip install -e .` but that might be as PyCharm is doing magic with adding content/source roots to sys.path. I'll check from a normal terminal. 

Also linking back to https://sourceforge.net/p/docutils/patches/186/ -- not sure if you've seen that discussion with Engelbert.

A

**Comment by AA-Turner** [comment]
(This in the changelog seems positive -- https://flit.readthedocs.io/en/latest/history.html#version-3-4 // https://github.com/pypa/flit/pull/400/)

**Comment by chrisjsewell** [comment]
Yeh I have cheers, but would rather not discuss things over sourceforge's horrible interface 😅

**Comment by AA-Turner** [comment]
Seems to be working!

```powershell
PS S:\Development\docutils-editable-test> python -V
Python 3.9.7
PS S:\Development\docutils-editable-test> python -m venv venv
PS S:\Development\docutils-editable-test> .\venv\Scripts\Activate.ps1
(venv) PS S:\Development\docutils-editable-test> python -m pip install -U --quiet pip
(venv) PS S:\Development\docutils-editable-test> python -m pip -V
pip 21.3.1 from S:\Development\docutils-editable-test\venv\lib\site-packages\pip (python 3.9)
(venv) PS S:\Development\docutils-editable-test> python -c "import docutils; print(docutils.__version__)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'docutils' has no attribute '__version__'
(venv) PS S:\Development\docutils-editable-test> python -m pip install --quiet -e docutils
(venv) PS S:\Development\docutils-editable-test> python -c "import docutils; print(docutils.__version__)"
0.19b.dev
```

A

**Comment by grubert** [comment]
strictly speaking : this patch changes the command line API, removing commands, so it can only be applied in a major version change, really ?
Removing commands will break a lot of scripts.
I understand https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html sees the same problem

> When you move away from direct setup.py invocations (note: this does not mean removing setup.py entirely or switching away from setuptools, it just means that you change all your scripts and documentation to tell people not to do this)

So my plan is to apply this piecewise, leave commands (the work to keep setup.* and pyproject.toml is mine anyway) AND keep the discussion assume to be nudged now and then.

**Comment by chrisjsewell** [comment]
> Seems to be working!

ah good to hear, I might give it a go again for some of my packages then 👍

**Comment by AA-Turner** [comment]
> this patch changes the command line API, removing commands

To make sure we're talking about the same thing here, do you mean removing `setup.py *` here?

I'm going to respond as if you do, but please correct me if I'm wrong.

If this requires a `0.20` instead of a `0.19.1` fair enough -- I think that's fine and within SemVer (depends on how widely you define breaking change, but if you include the build system in that definition then fair enough).

The overwhelming majority of the users of Docutils will get it via a pre-built wheel though, so they won't experience any issues.

> my plan is to apply this piecewise

Could you advise on your thinking here? (i.e. which bits are you going to defer / apply now)

I'd be very happy to prepare a smaller patch, but I don't fully understand what you're proposing, sorry.

A

**Comment by AA-Turner** [comment]
comment in reply to Günter: https://sourceforge.net/p/docutils/patches/186/#897a

**Comment by AA-Turner** [comment]
PR updated to use setup.cfg-based declarative metadata (less good than pyproject.toml based, but a smaller change for the maintainers)

xref https://sourceforge.net/p/docutils/patches/186/#897a/547e

A

**Comment by AA-Turner** [comment]
Rebased onto latest `master`

File: docutils/docs/dev/release.txt [hunk 1]
```diff
@@ -40,8 +40,8 @@ Releasing (post 2020)
 
 * Generate universal wheel and source-distribution::
```
Comment (hugovk) [inline docutils/docs/dev/release.txt:41]:
```suggestion
* Generate wheel and source-distribution::
```

File: docutils/docs/dev/release.txt [hunk 2]
```diff
@@ -40,8 +40,8 @@ Releasing (post 2020)
...
 
-    python3 setup.py sdist
-    python3 setup.py bdist_wheel --universal
+    python3 -m pip install build
+    python3 -m build .
 
 * Upload universal wheel and source to test.pypi::
```
Comment (hugovk) [inline docutils/docs/dev/release.txt:46]:
```suggestion
* Upload wheel and source to test.pypi::
```

File: docutils/docs/dev/release.txt [hunk 3]
```diff
@@ -71,8 +71,8 @@ Releasing (post 2020)
 * Update your source directory. 
 * Rebuild universal wheel and source-distribution ::
```
Comment (hugovk) [inline docutils/docs/dev/release.txt:72]:
```suggestion
* Rebuild wheel and source-distribution ::
```

**Review by hugovk — COMMENTED** [review]
We can update these comments too.

**Comment by hugovk** [comment]
> If this requires a `0.20` instead of a `0.19.1` fair enough -- I think that's fine and within SemVer (depends on how widely you define breaking change, but if you include the build system in that definition then fair enough).

Technically, it's totally within SemVer: for 0.x anything goes. https://semver.org/ says:

> 4. Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

**Comment by AA-Turner** [comment]
From Günter in [patches#186](https://sourceforge.net/p/docutils/patches/186/?page=2#621a):

> The generic front end is available via `python -m docutils` and `docutils` CLI commands since [[r9061]](https://sourceforge.net/p/docutils/code/9061/).
> 
> Updated timeline suggestion:
> 
> * release 0.19
> * Decide, which `rst2*` entry points we want to keep (all, the most used, ...).
> * Provide "entry points" for them in `setup.py` (-> command line commands without `.py` extension in Unix, Mac, and Windows).
> * Add deprecation notice in RELEASE-NOTES and docs/usr/tools. Change usage examples in the documentation.
> * Announce deprecation of rst2*.py commands at runtime (the deprecation message contains the replacement invocation needed. This may be a `docutils` command with options or one of the abovementioned new entry points). The replacement must be implementd at the time we start to nag users.
> * release 1.1
> * In the 2.0 development branch: Cease installing rst2*.py tools in the PATH, allows switch to `flit` and `*.toml` config file.
> * Eventually provide a separate "docutils-cli" package which may include moving the generic `docutils` front-end tool there and/or providing the special tools as `rst2*` or `rst2*.py`.
> * release 2.0 with modern packaging.

So this patch won't be fully applied before Docutils 2.0 is released. (Note #6 is actually more up to date than this, but neither would be applied as I'll need to make a new patch etc at each of the steps Günter outlined implementing the changes.

A

**Pull request closed without merging**